Add monad laws

Kha opened the pull request:
* add invariants to `functor`, `applicative`, and `monad`
* default as many superclass laws as possible - when given only `pure` and `bind`, the proof obligations break down to just 1 functor law and 2 monad laws
* prove non-meta instances correct, use `undefined` for meta ones
* extend `structure` a bit to make things easier

The instance proofs aren't exactly pretty - I wildly switched between `dsimp`, `simp`, and `rw` depending on what worked in a given context.

I tried to fix `io` by constructing an unsafe `monad` instance in C++ via Lean, but there's still some VM error left. There are also a few other tests I have broken somehow.

Kha committed:
feat(frontends/lean/util): allow lazily resolved auto params

Kha committed:
fix(frontends/lean/elaborator): elaborate auto_param fields last

Kha committed:
feat(init): add default value proofs to the monadic hierarchy

Kha committed:
feat(init/category/monad): simplify `monad.seq._default`

Kha committed:
feat(library): add functor, applicative, and monad laws, and prove them correct for non-meta instances

Kha committed:
feat(frontends/lean/elaborator): elaborate auto params and explicitly given fields as soon as possible, too.

Also make sure not to elaborate fields before other fields their type depends on.
Makes inline tactic proofs in structure instances possible.

Kha committed:
feat(frontends/lean/structure_cmd): allow local notations

Kha committed:
feat(library/tactic/elaborator_exception): show context of failed placeholders

Kha committed:
fix(system/io): try to fix io monad

leodemoura commented:
> I tried to fix io by constructing an unsafe monad instance in C++ via Lean, but there's still some VM error left. There are also a few other tests I have broken somehow.

I will take a look.

leodemoura committed:
fix(library/vm/vm_io): fix io_monad

see #1485

leodemoura commented:
Fixed the io_monad

leodemoura commented:
> The instance proofs aren't exactly pretty - I wildly switched between dsimp, simp, and rw depending on what worked in a given context.

I investigated some of your proofs. I can see it was not a smooth ride.
They expose problems with `simp`. 
I found the source of the problem (higher order matching), but it is not an easy fix. 
I will create a new wiki for describing a new matching algorithm for Lean.

leodemoura committed:
fix(frontends/lean/elaborator): ignore auto/default params when `@` is used

see #1485

@kha We need this commit to be able to execute commands such as:
```lean
  #check @monad.mk
```

leodemoura commented on library/init/data/option_t.lean:
```diff
@@ -0,0 +1,71 @@
+   cases x,
+   { dsimp [option_t_bind._match_1, return, pure],
+     rw [monad.pure_bind], apply rfl },
+   { apply rfl }
```
We have `tactic.interactive.refl`.

leodemoura committed:
feat(library/init/meta/tactic): add apply_auto_param and apply_opt_param tactics

see #1485

Kha committed:
fix(init/meta/interactive): mk_simp_set: also remove equational lemmas

Kha committed:
refactor(init/data/list/instances): simplify proofs

Kha committed:
fix(library/unfold_macros): recursively unfold macros

leodemoura commented on src/library/unfold_macros.cpp:
```diff
@@ -29,7 +29,7 @@ class unfold_untrusted_macros_fn : public replace_visitor_with_tc {
             if (optional<expr> new_r = m_ctx.expand_macro(r)) {
-                return *new_r;
+                return visit(*new_r);
             } else {
```
BTW, this module will need to be re-implemented. It was already a performance bottleneck before this change.

Kha commented:
> Fixed the io_monad

Thanks! There is a last trust0 issue with macros in intro rules not being expanded, but I'm not sure where to best place a fix.

leodemoura commented:
> There is a last trust0 issue with macros in intro rules not being expanded, but I'm not sure where to best place a fix.

Did the change at `unfold_macros` fixed the problem?
If yes, then keep it. 
We will need to re-implement `unfold_macros` anyway in the future. 
We already have performance problems in this module. 
If the change did not fix the problem, then try to look for missing `unfold_untrusted_macros` calls at `structure_cmd.cpp`.

Kha commented:
Sorry, I should have been more specific. I was looking for a catch-all location to place the call at, but I don't think there is one in library/. I'll add one just in structure_cmd for now.

leodemoura commented:
@kha I have been trying to implement the "abstract nested proofs" feature. It was much harder than expected. I found many problems. I can list them if you are interested. Here is the compromise I implemented.
 
Proofs are abstracted in 

1) Regular definitions (i.e., no recursive equation) `def f : T := v`, where `T` is not a proposition, and it is not meta.

2) Right hand side of equation lemmas (if its type is not a proposition). We use these equation lemmas for "unfolding" the definition. Example: 
```lean
#check  @vector.append.equations._eqn_1
```
produces
```lean
∀ {α} {n m} l₁ h₁ l₂ h₂,  append ⟨l₁, h₁⟩ ⟨l₂, h₂⟩ = ⟨l₁ ++ l₂, append._main._proof_1 l₁ h₁ l₂ h₂⟩
```
Before this change, we would get
```lean
∀ {α} {n m} l₁ h₁ l₂ h₂,  append ⟨l₁, h₁⟩ ⟨l₂, h₂⟩ = ⟨l₁ ++ l₂, ... huge mess ...⟩
```

Kha committed:
fix(frontends/lean/structure_cmd): unfold untrusted macros in intro rules

Kha committed:
chore(.appveyor.yml,.travis.yml): invoke ctest with --output-on-failure

Kha committed:
fix(library/tactic/eval,kernel/kernel_exception): hide internal definition name on eval_expr type error

cipher1024 commented:
In terms of design, is there a reason to have the functor, applicative and monad methods provided in the same classes as the laws? I find the layering of orders (i.e. has_le, weak_order, linear_weak_order, strict_order, etc) and arithmetic (has_add, add_semigroup, add_monoid, etc) very appealing and it makes them easier to use.

leodemoura commented:
@cipher1024 Are you suggesting we start with simple classes such as `has_bind`, `has_map`, `has_seq`, ... (similar to `has_add`, `has_mul`), and then we define `monad`, `functor`, etc on top of them?

BTW, we are planning to revise the algebraic hierarchy. If you have time, it would be great to have your feedback. The proposal is here: https://github.com/leanprover/lean/wiki/Refactoring-structures

cipher1024 commented:
@leodemoura I was mostly referring to the separation that exists in the Lean library between the type classes with functions and the type classes with laws. I find it especially useful because, typically, when you define functions in terms of monads, what you really need is to have access to the bind operator, regardless of how it behaves. Later, when you prove something about the function, then you should assume that bind is associative, etc. 

This being said, I like the idea of having the basic `has_bind`, `has_map`, `has_seq`, `has_pure` but only in so far as every new operator adds expressive power. For instance, I don't think that we should have `has_seq_left` or `has_seq_right`. I would rather make them a part of `has_seq`.

I'm thinking that the classes with laws could follow the hierarchy of the [semigroupoids](https://hackage.haskell.org/package/semigroupoids-5.1/docs/Data-Functor-Bind.html) Haskell library with respect to Functor, Apply, Applicative, Bind and Monad. 

BTW, thanks for linking me to the refactoring project. It looks exciting :) I'll contribute whatever feedback I can.

Kha committed:
fix(library/module): unfold untrusted macros in imported intro rules

Kha commented:
@leodemoura
> Proofs are abstracted in
> 
>     Regular definitions (i.e., no recursive equation) def f : T := v, where T is not a proposition, and it is not meta.
> 
>     Right hand side of equation lemmas (if its type is not a proposition). We use these equation lemmas for "unfolding" the definition.

Nice, that sounds good enough. For default proofs in class decls, the `_default` definitions already act as abstraction, right?

Kha commented:
@cipher1024
> This being said, I like the idea of having the basic has_bind, has_map, has_seq, has_pure but only in so far as every new operator adds expressive power. For instance, I don't think that we should have has_seq_left or has_seq_right. I would rather make them a part of has_seq.

Sure, we can always add those (note that we already have `has_bind`). But I think it's a good idea that something called `monad` also has to fulfill the monad laws.

> I'm thinking that the classes with laws could follow the hierarchy of the semigroupoids Haskell library with respect to Functor, Apply, Applicative, Bind and Monad.

Feel free to insert the missing classes into the hierarchy. I also haven't added laws to `applicative` yet, though it's not [quite clear](https://wiki.haskell.org/Typeclassopedia#Laws_6) what those should be.

spl commented:
@Kha:

> I also haven't added laws to `applicative` yet, though it's not quite clear what those should be.

I think you meant `alternative` there. :)

Kha commented:
> I think you meant alternative there. :)

Oh, right :) .

johoelzl commented:
@Kha it looks like we need a hierarchy of `alternative` type classes. What the Typeclassopedia article does not mention: how should `map` behave under `<|>`. (a weaker form of right distributivity) I would assume it distributes nicely `map f (a <|> b) = (map f a) <|> (map f b)`. The problems mentioned in the article go away when the first part of `seq` is `pure`.

leodemoura commented:
> Nice, that sounds good enough. For default proofs in class decls, the _default definitions already act as abstraction, right?

@kha Yes, this is correct.
